### DISCUSSION TITLE
GSoC 2023 Project Ideas

### DISCUSSION THREAD

**Discussion by mwengren** [ioos/nextgen-dmac#34] [General]
From @rsignell-usgs:

Micah,

Thanks for posting about this opportunity.    It was a great experience for me to serve as a co-mentor for Summer of Code student Peter Marsh last year, and Peter was super productive, making new workflows, uncovering bugs and potential enhancements to Kerchunk which makes accessing IOOS model output more cloud-performant!   (for more info see his blog post: [Accessing NetCDF and GRIB file collections as cloud-native virtual datasets using Kerchunk | by Peter Marsh | pangeo | Medium](https://medium.com/pangeo/accessing-netcdf-and-grib-file-collections-as-cloud-native-virtual-datasets-using-kerchunk-625a2d0a9191))

I'm cc'ing @abkfenris on this chain also because he has a cool idea for using kerchunk and Xarray's DataTree to create virtual Forecast Model Run Collections (FRMC for you THREDDS fans) of the type we saw David Stuebe demo last time on the Committee Call. 

--
Original request from @mwengren:

IOOS is applying to Google Summer of Code 2023.  IOOS was accepted as a mentoring organization for both GSoC 2021 and 2022, with many open source project successes as a result:

2022: https://github.com/ioos/gsoc/blob/main/2022/project-results.md 
2021: https://github.com/ioos/gsoc/blob/main/2021/project-results.md 

The application deadline this year is Feb 7.  In order to be re-accepted as a mentoring organization, we need as many project ideas submitted as possible by this deadline. 

We need your help!  If you are interested in mentoring a qualified student over the summer to contribute code, documentation, training materials, or use case examples to an ocean data-related open source software project you’re involved with, please go to the IOOS Google Summer of Code GitHub repository (https://github.com/ioos/gsoc) and follow links there to submit your project idea.

Project ideas need to include basic background information (project description and purpose, expected skills, potential mentors), but it’s a fairly easy lift to propose a new project.  

For more information on the GSoC program, timeline, and expectations of potential mentors, follow this link: https://summerofcode.withgoogle.com/how-it-works.  

Mentors are expected to be able to commit a few hours per week working with students on their contributions: reviewing submitted code/documentation and providing feedback to students to ensure a successful outcome for both students and mentors.  

Note: while students who pass evaluation at the end of the program will receive a stipend from Google for their work, IOOS must decline all mentor stipend payments from Google.  The experience of helping a student gain valuable exposure to open source development and furthering your own projects is reward enough!

If you’re on the fence, please reach out to any of IOOS’ prior GSoC mentors by looking at the project results pages linked above, or visiting this short list of [past projects](https://github.com/ioos/gsoc/issues?q=is%3Aissue+label%3AGSoC%2CGSoC21%2CGSoC22).

IOOS will collect your submitted project ideas and add them in our 2023 ideas list for review by Google on Feb 7: https://github.com/ioos/gsoc/blob/main/2023/ideas-list.md.  

Please reach out to Micah ([micah.wengren@noaa.gov](mailto:micah.wengren@noaa.gov)), Matt ([Mathew.Biddle@noaa.gov](mailto:Mathew.Biddle@noaa.gov)), or Filipe ([ocefpaf@gmail.com](mailto:ocefpaf@gmail.com)) with any questions.

**Comment by mwengren** [comment]
Hi all - just a nudge that we have less than a week to go for input on project ideas for GSOC 23.

We don't need much detail for any potential projects, just answers to the questions in [this form](https://github.com/ioos/gsoc/issues/new?assignees=&labels=GSoC23%2C+project+idea&template=gsoc-project-proposal.md&title=Project+title).

@mpiannucci Is there something at an introductory/intermediate python level that a student could tackle related to the xreds work you're doing?
@abkfenris You were involved last year with the erddapy project with @ocefpaf.  Based on that experience, do you think a XPublish-related project would be fruitful?   Do you have bandwidth to mentor?
@jonmjoyce data ingest is an area we haven't talked much about yet, any potential projects there?
@rsignell-usgs I know you're scaling back, wondering if you have any suggestions though?  What about guiding a student through more kerchunk workflows for some of the datasets that are within our scope here?  We can provide access to whatever cloud resources we have as part of this project for them to crunch through converting some datasets/generating some kerchunk metadata outputs.  I'm not sure what the in scope datasets are exactly, or where the cloud resources are, but hopefully those questions can be answered...

If we end up with some good project possibilities, we can try to get word out to get student interest via the [Pangeo Discourse](https://discourse.pangeo.io/) forum.

**Reply by abkfenris** [reply]
Ya, I think something Xpublish /XREDs related could be a good project, but I'm kinda hung up on what the right scale is, as I think we overshot last year. [We just landed the plugin system](https://github.com/xarray-contrib/xpublish/pull/146) (and I got commit bit), so we now have the groundwork to build on. 

I think I've got the bandwidth to mentor, especially if we can find someone who is willing to turn learning into documentation (I teach them things that are unclear, and then they can document them).

Some roughly small > large ideas:
- STAC Catalog dataset loading plugin
- Extend my OGC EDR router to support other query geometries
- IOOS QC validation router
- - Extend OGC EDR router for format support and write a ERDDAP API compatible router to wrap it
- THREDDS catalog.xml dataset loader
- ERDDAP datasets.xml dataset loader
- Xpublish CLI and config handling
- Work on @mpiannucci WMS router and similar tile/image routing. Maybe refactor to use more of rio-tiler under the hood now that it has direct Xarray support.
- Xpublish tabular data support
- Xpublish datatree support
  - Forecast support with Xarray-FMRC and re-using dataset routers for forecasts. May also have a kerchunk element
- Xpublish UI plugin infrastructure (designing a standard browser user focused interface, and how can other plugins contribute elements/pages to it)

**Comment by mwengren** [comment]
@abkfenris that looks like a great list!  I think as long as there's a technical coding component, a project that's part code and part documenting approaches for building out a project/API is within scope.  Last year's Kerchunk project was applying the kerchunk approach to a particular use case/dataset, documenting it for others, and also contributing improvements to kerchunk at the same time, and that fit in quite well. 

There's also the [Google Season of Docs](https://developers.google.com/season-of-docs) (which we don't participate in), but that's really geared to writers and not developers.  You're talking about teaching a developer how to build something new and document it well in the process.  If you want to submit multiple ideas at this point, that's fine.  Ultimately, you'll have to pick the one that you think has the best chance of success after you get student proposals back.

**Comment by mwengren** [comment]
Hi all, we only have a day left to submit project ideas by the Feb 7 deadline for IOOS to apply for GSoC 23.

I heard @abkfenris wasn't going to be able to mentor this year, so I'm wondering if any of the other ideas germinating here for the Reaching for the Cloud project could be written up as a short project proposal.  

Does anyone else have interest in mentoring someone for some of this work?  It could be a great opportunity to allow someone else to learn from experience with a lot of these tools, while also getting some things accomplished to move this project forward.

**Reply by jonmjoyce** [reply]
@mwengren we've discussed the GSoC opportunity internally and weren't able to come up with something that wouldn't distract from our prototyping goals. We also have already engaged in a different opportunity to work with students this summer which will obviously require its own mentoring overhead. I think this is a good opportunity but we just don't really have the bandwidth to engage in it this year.
